src/irc.py

```diff
@@ -129,52 +127,121 @@ def _install_snap_package(
             logger.exception(error_msg)
             raise InstallError(error_msg) from e
 
-    def _write(self, path: pathlib.Path, source: str) -> None:
-        """Pushes a file to the unit.
+    def configure(
+        self, db: DatasourcePostgreSQL, matrix: DatasourceMatrix, config: CharmConfig
+    ) -> None:
+        """Configure the service.
 
         Args:
-            path: The path of the file
-            source: The contents of the file to be pushed
+            db: the database configuration
+            matrix: the matrix configuration
+            config: the charm configuration
         """
-        path.write_text(source, encoding="utf-8")
-        logger.info("Pushed file %s", path)
-
-    def _generate_PEM_file_local(self) -> str:
-        """Generate the PEM file content.
+        self._generate_pem_file_local()
+        self._generate_app_registration_local(matrix, config)
+        self._eval_conf_local(db, matrix, config)
+
+    def _generate_pem_file_local(self) -> None:
+        """Generate the PEM file content."""
+        pem_create_command = [
+            "/bin/bash",
+            "-c",
+            f"[[ -f {IRC_BRIDGE_PEM_FILE_PATH} ]] || "
+            f"openssl genpkey -out {IRC_BRIDGE_PEM_FILE_PATH} "
+            f"-outform PEM -algorithm {IRC_BRIDGE_KEY_ALGO} -pkeyopt {IRC_BRIDGE_KEY_OPTS}",
+        ]
+        logger.info("Creating PEM file for IRC bridge.")
+        subprocess.run(pem_create_command, shell=True, check=True, capture_output=True)  # nosec
+
+    def _generate_app_registration_local(
+        self, matrix: DatasourceMatrix, config: CharmConfig
+    ) -> None:
+        """Generate the content of the app registration file.
 
-        Returns:
-            A string
+        Args:
+            matrix: the matrix configuration
+            config: the charm configuration
         """
-        pass
-
-    def _generate_conf_local(self) -> str:
+        app_reg_create_command = [
+            "/bin/bash",
+            "-c",
+            f"[[ -f {IRC_BRIDGE_REGISTRATION_FILE_PATH} ]] || "
+            f"matrix-appservice-irc -r -f {IRC_BRIDGE_REGISTRATION_FILE_PATH}"
+            f" -u https://{matrix.host}:{IRC_BRIDGE_HEALTH_PORT} "
```

Comment:
How do we intend to route traffic to the bridge machine ?
If they share the same host as matrix, then we need a reverse proxy in front of both that will route based on path or port since the hostname will resolve to the same IP for both synapse and the bridge

Reply:
Well spotted, this is not correct, it should be the address where the Matrix homeserver can contact the bridge. It can be a config option if we use an FQDN or an IP that is reachable from Matrix, wdyt?

Reply:
Yeap that's tricky... the machine needs either to be accessible publicly or internally, and it should support HTTPS meaning you need either a TLS termination or support the TLS relation in the bridge :\

Reply:
Depends on the communication flow, I think Matrix and the bridge talk to each other, soo it could be private? And the public part between the IRC network and the bridge happens over the Mattix homeserver. Need to check, you raised a very good point here.